forest/tree/tree_writer.go

```diff
@@ -0,0 +1,128 @@
+package tree
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"hash/fnv"
+	"io/ioutil"
+	"sort"
+
+	"github.com/treeverse/lakefs/forest/sstable"
+	gr "github.com/treeverse/lakefs/graveler"
+)
+
+const (
+	SplitFactor    = 200_000 // avarage number of entries in a part file. used to compute modulo on hash of path
+	SplitMaxfactor = 3       // a part will be closed if number of entries written to it exceed splitFactor * splitMaxFactor
+	SplitMinFactor = 50      // a part will not be closed in number of rows less than splitFactor / SplitMinFactor
+)
+
+type IsSplitKeyFunc func(path gr.Key, rowNum int) bool
+
+type TreeWriter struct {
```

Comment:
I think the TreeWriter is coupled to `Apply` more than we want it to be.
Some of the inner logic (like flushIterToNewTree) is fine for `Apply`, but might differ for `Merge` for example.